imgurbash2

```diff
@@ -22,216 +22,599 @@
-	echo "" >&2
-	echo "Where:" >&2
-	echo "    image_num is a local image file or a remote image (i.e. http/https)." >&2
-	echo "" >&2
-	echo "Options:" >&2
-	echo "    -d, --delete <delete_hash_1> [<delete_hash2>...]" >&2
-	echo "        Deletes image(s) based on the the supplied delete hash.  The delete hash" >&2
-	echo "        is outputted when the user uploads an image." >&2
-	echo "    -h, --help" >&2
-	echo "        Displays this help and exit." >&2
-	echo "    -v, --version" >&2
-	echo "        Print program version and exit." >&2
-	echo "" >&2
-	echo "Examples:" >&2
-	echo "    Refer to https://github.com/ram-on/imgurbash2" >&2
+# @returns {bool}  true, if provided url was a valid url.
+is_valid_url() {
+    local url regex
+
+    url="$1"
+
+    regex='https?://[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]'
+
+    [[ "$url" =~ $regex ]]
+}
+
+
+# Checks whether given string is empty or 'null'.
+#
+# @param {string}  str   string to check.
+#
+# @returns {bool}  true, if provided string is empty or 'null'
+is_null_or_empty() {
+    [[ -z "$*" || "$*" == null ]]
 }
 
 
+# Performs call to imgur api.
+# Injects appropriate curl headers, depending on whether authorized
+# or anon request is being made.
 #
-# Returns true if the image is stored on a remote machine (i.e. http or https);
-# false otherwise.
+# @param {string}     -r resource   api resource to hit.
+# @param {string...}  hdrs          additional curl options.
+#
+# @returns {string}  response body, as returned by curl.
+api_call() {
+    local hdrs resource
+
+    if [[ "$1" == '-r' || "$1" == '--resource' ]] && [[ -n "$2" ]]; then
+        resource="$2"
+        shift 2
+    else
+        fail "no resource given when calling ${FUNCNAME}()"
+    fi
+
+    declare -a hdrs=("$@")
+    [[ "$LOGIN" == true ]] && hdrs+=('-H' "Authorization: Bearer ${ACCESS_TOKEN}") \
+            || hdrs+=('-H' "Authorization: Client-ID $ANON_CLIENT_ID")
+
+    curl --compressed \
+            --connect-timeout "$UPLOAD_CONNECT_TIMEOUT" \
+            -m "$UPLOAD_TIMEOUT" \
+            --retry "$UPLOAD_RETRIES" \
+            --stderr - \
+            -sSL \
+            "${hdrs[@]}" \
+            "https://api.imgur.com/3/${resource}"
+}
+
+
+# Uploads given image to imgur.
 #
-# Parameter:  URL to parse
+# @param {string}  file   path to local image file, or external URL.
 #
-function is_remote_image() {
-	if [ $(grep -i -E "^(http[s]?)://.+" <<< "$1")  ]; then
-		return $(true)
-	else
-		return $(false)
-	fi
+# @returns {string}  response body, as returned by curl.
+upload_image() {
+    local file hdrs
+
+    file="$1"
+    declare -a hdrs=('-X' POST)
+
+    [[ -n "$TITLE" ]] && hdrs+=('-F' "title=$TITLE")
+    [[ -n "$ALBUM_ID" ]] && hdrs+=('-F' "album=$ALBUM_ID")
+
+    if [[ -f "$file" ]]; then
+        hdrs+=('-F' "image=@$file")
+    elif is_valid_url "$file"; then
+        # TODO: only works when album_id not defined?
```

Comment:
Could anyone confirm logic for external image upload is valid?
For some reason can't get it to work if we want to upload to a specific album.